consensus/src/consensus/request_response.rs

```diff
@@ -22,45 +27,109 @@ impl<N: Network> Consensus<N> {
+            blockchain,
+            policy::MAX_RESPONSE_REQUEST_BLOCK,
+            policy::MAX_RESPONSE_REQUEST_BLOCK_RANGE,
+        ));
 
         let stream = network.receive_requests::<RequestMissingBlocks>();
-        tokio::spawn(Self::request_handler(network, stream, blockchain));
+        tokio::spawn(Self::request_handler(
+            network,
+            stream,
+            blockchain,
+            policy::MAX_RESPONSE_REQUEST_MISSING_BLOCKS,
+            policy::MAX_RESPONSE_REQUEST_BLOCK_RANGE,
+        ));
 
         let stream = network.receive_requests::<RequestHead>();
-        tokio::spawn(Self::request_handler(network, stream, blockchain));
+        tokio::spawn(Self::request_handler(
+            network,
+            stream,
+            blockchain,
+            policy::MAX_RESPONSE_REQUEST_HEAD,
+            policy::MAX_RESPONSE_REQUEST_BLOCK_RANGE,
+        ));
     }
 
-    pub(crate) fn request_handler<Req: Handle<Req::Response> + Request>(
+    pub(crate) fn request_handler<Req: Handle<Req::Response, N> + Request>(
         network: &Arc<N>,
         stream: BoxStream<'static, (Req, N::RequestId, N::PeerId)>,
         blockchain: &Arc<RwLock<Blockchain>>,
+        max_requests: u16,
+        block_range: u32,
     ) -> impl Future<Output = ()> {
         let blockchain = Arc::clone(blockchain);
         let network = Arc::clone(network);
+        let peers_request_limit = Arc::new(futures::lock::Mutex::new(HashMap::with_capacity(
+            network.get_peers().len() * 5,
+        )));
+
         async move {
             stream
                 .for_each_concurrent(
                     Self::MAX_CONCURRENT_HANDLERS,
                     |(msg, request_id, peer_id)| {
-                        let request_id = request_id;
                         let network = Arc::clone(&network);
                         let blockchain = Arc::clone(&blockchain);
+                        let peers_request_limit = Arc::clone(&peers_request_limit);
                         async move {
-                            let blockchain = Arc::clone(&blockchain);
-                            let network = Arc::clone(&network);
-                            let request_id = request_id;
                             tokio::spawn(async move {
-                                trace!("[{:?}] {:?} {:#?}", request_id, peer_id, msg);
+                               trace!("[{:?}] {:?} {:#?}", request_id, peer_id, msg);
+
+                                // Gets lock of peer requests limits read and write on it
+                                let mut peers_request_limit = peers_request_limit.lock().await;
+                                // If the peer has never sent a request of this type, creates a new entry
+                                let requests_limit = peers_request_limit
+                                    .entry(peer_id)
+                                    .or_insert_with(|| RateLimit::new(max_requests, block_range));
```

Comment:
These entries never get evicted, letting this Map grow forever.

We need to introduce means to remove peers from that map once we are no longer connected to them. Likely it would also be necessary to wait with the removal until the `last_reset + block_range` block was pushed as one could otherwise bypass the rate limiting by disconnecting and reconnecting.

Reply:
This will be complicated to implement i believe. Basically, the rate-limiting is being implemented where the processing of the requests is at, which is out of the network.  The network is where the disconnect peer function exists.